[Sean Tan Rui An] iP

thatseant opened the pull request.

thatseant committed:
Add welcome messages (Level-0)

Add welcome message and goodbye message before exiting.

Duke exits before responding to any user input.

thatseant committed:
Echoes commands entered by user

thatseant committed:
Stores and displays user inputs when requested

thatseant committed:
Add task completion feature

Done command followed by task number marks particular task as completed. List command now shows task status as well.

User inputs previously stored in String Array.

They are now stored as Tasks (new Task Class created). Tasks track completion status.

thatseant committed:
Extracts commands implementation as methods

displayList, markAsDone and addTask are new methods that implements Duke’s commands, previously done in main().

numberOfTasks now stored as class-level attribute in Tasks and not in Duke class.

toString method in Task Class prints status of task together with name.

thatseant committed:
Classes Event, Deadline, Todo inherits from Task

addTask function in Duke takes note of taskType and creates corresponding Event, Deadline or Todo object.

toString method in classes overridden to include task type indicator and event date or deadline.

samuellleow commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,75 @@
+        String userInput = in.nextLine();
+
+        while (!userInput.equals("bye")) {
+            if (userInput.equals("list")) {
+                displayList();
+            } else if (userInput.contains("done")) {
+                int taskNumber = Integer.parseInt(userInput.replaceAll("\\D+","")) - 1;
+                markAsDone(taskNumber);
+            } else {
+                addTask(userInput);
```
I like the use of extraction to determine the task type, display list and task done function. It makes the code more concise and easy to read.

thatseant replied:
Thanks Samuel!! I'm not sure if I should abstract it more, ie create a commands-handling function that calls the three functions. Would it make the code harder to read?

samuellleow reviewed (commented):
Looks good.
I feel that the code is easy to read for most parts.

thatseant committed:
Added Test Cases

Move Task and its child classes to Task folder.

Added test cases in input.txt and ensures runtest.sh compiles with Task classes too.

Fix bug where event and deadline “by/at” only reads single word.

thatseant committed:
Ensures SLAP and remove Magic Numbers

SLAP - Command processing loop in Duke now uses Switch statement. Each case in Switch statment only does calling of functions.

Remove Magic Numbers - Tasks array capacity defined as TASKS_CAPACITY.

Regex in string splitting simplified.

Test cases expected output corrected.

thatseant committed:
Fix bug where task name has an additional space

shreytheshreyas commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,87 @@
+import Task.Task;
```
Hi Sean, Good Job on your code I found no violation to the coding standards in your code. Just one comment  i don't think you have to expliciltly import your classes because they are all kept in the same file other than that everything seems great.

thatseant replied:
Hi they're in different files in the same package. Was told we had to import individually.

0xEljh commented on src/main/java/Duke.java:
```diff
@@ -16,13 +16,30 @@ public static void displayList() {
+            break;
+        }
+        System.out.format("Got it. I've added this task: %n");
+        System.out.format("%s%n", tasks[Task.getNumberOfTasks()-1].toString());
```
Consider leaving spaces between operators for better readability

thatseant replied:
Hi which operators you referring to?

0xEljh commented on src/main/java/Duke.java:
```diff
@@ -1,5 +1,30 @@
+    static Scanner in = new Scanner(System.in);
+    static Task[] tasks = new Task[100];
+
+    //Prints tasks with completion status
```
Consider the use of javadocs to describe the behavior of your methods.

0xEljh reviewed (commented):
Overall, your code has good readability standards. The main comment I have for you would be to consider converting your comments on methods where you use in-line comments into javadocs/multi-line comments.

thatseant committed:
Added Error Handling

Handles incorrect inputs entered by user.

Incorrect inputs handled via try/catch:
1. Missing task name/description
2. Deadline without deadline specified
3. Event without event time
4. Invalid task number for task completion
5. Invalid Command (by throwing new Class InvalidCommandException)

thatseant committed:
Added more test cases

Tests now handles exception.

thatseant committed:
Merge branch 'branch-Level-5'

thatseant committed:
Stores task in ArrayList instead of Array

thatseant committed:
Added Delete Function

Implemented in deleteTask.

To work on: Repetitive code in printing status of task similar to markAsDone().

thatseant committed:
Saves and Loads Tasks

Saves to data file by overwriting after every user commands. Loads data at start of program into Tasks array.

Save: saveToFile() and writeToFile()
Load: readFile() and loadFromFile()

thatseant committed:
Merge branch 'branch-Level-6'

thatseant committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Duke.java

okkhoy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,208 @@
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.Scanner;
 public class Duke {
+
```
Could use some vertical spacing for readability

okkhoy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,208 @@
+
+    public static void saveToFile() {
+        String fullTextToAdd = "";
+        try {
+            for (Task task: tasks) {
+                if (task != null) {
+                    fullTextToAdd += task.toRawData();
+                    fullTextToAdd += System.lineSeparator();
+                }
+            }
+            writeToFile(filePath, fullTextToAdd);
+        } catch (IOException e) {
```
Try to make the happy path prominent

thatseant replied:
What do you mean by prominent? The only edge case is if task==null and I shouldn't flip this if statement right?

okkhoy replied:
yes. handle the edge case first. then automatically, happy path becomes the main thing to see in the method.

okkhoy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,208 @@
+    public static void loadFromFile() throws FileNotFoundException {
+        File f = new File(filePath); // create a File for the given file path
+        Scanner s = new Scanner(f); // create a Scanner using the File as the source
+        while (s.hasNext()) {
+                String[] taskParameters = s.nextLine().split(" \\| ");
+                String taskType = taskParameters[0];
+                Boolean isDone = Boolean.parseBoolean(taskParameters[1].trim());
+                String taskName = taskParameters[2];
+
+                switch (taskType) {
+                case "T":
+                    tasks.add(new Todo(taskName));
+                    break;
+                case "D":
+                    String date = taskParameters[3];
+                    tasks.add(new Deadline(taskName, date));
+                    break;
+                case "E":
+                    date = taskParameters[3];
+                    tasks.add(new Event(taskName, date));
+                    break;
+                default:
+                    break;
+                }
+                if (isDone) {
+                    tasks.get(Task.getNumberOfTasks()-1).setDone();
+                }
+        }
```
Looks like indentation is off here.
Tip: How would you get the IDE to fix this in one-shot?

thatseant replied:
Thanks! I've use IntelliJ reformat code.

okkhoy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,208 @@
+            try {
+                String commandType = userInput.split(" ")[0];
+                switch (commandType) {
+                case "list":
```
Consider using named constants in place of the command words.

thatseant replied:
Thanks for the suggestion. I always thought named constants were only to avoid magic numbers and constants that are unclear. What's the purpose of them here or is it just for convention?

okkhoy replied:
the command words are also _sort of_ magic literals, no?

okkhoy commented on src/main/java/Exception/InvalidCommandException.java:
```diff
@@ -0,0 +1,5 @@
+package Exception;
+
+public class InvalidCommandException extends Exception {
+
```
Here, use the opportunity provided by inheritance to include a more expressive error message rather than just create an empty class

okkhoy reviewed (commented):
Good job on following the coding standards!

thatseant commented:
> Good job on following the coding standards!

Thanks Prof for looking through and giving me targeted feedback. Appreciate it!

thatseant committed:
Auto reformat code in IDE

Code reformatted by Intellij

thatseant committed:
Refactor code for more OOP

Extracted Ui, Storage, Parser and Command classes.

thatseant committed:
Recognises deadline dates in yyyy-mm-dd format

Prints in MMM dd yyyy format.

thatseant committed:
Find Command allows users to search for tasks.

Search terms can include multiple words.

thatseant committed:
Add JavaDoc comments for all public methods

thatseant committed:
Merge pull request #1 from thatseant/branch-Level-8

Recognises deadline dates in yyyy-mm-dd format

thatseant committed:
Merge branch 'master' into branch-Level-9

thatseant committed:
Merge pull request #2 from thatseant/branch-Level-9

Find Command allows users to search for tasks.

thatseant committed:
Merge branch 'master' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/Command/DoneCommand.java
#	src/main/java/Task/Deadline.java

thatseant committed:
Merge pull request #3 from thatseant/branch-A-JavaDoc

Add header comments for all public methods and classes

thatseant committed:
Recognises deadline not in yyyy-mm-dd

Fixed bug in constructor to check if deadline is in yyyy-mm-dd format.

thatseant committed:
Declare constants and include exception messages

thatseant committed:
Updated user guide

User guide describes all commands.

thatseant committed:
Update README.md

thatseant committed:
Set theme jekyll-theme-tactile

thatseant committed:
Fix IndexOutOfBoundsException and New Test Cases

thatseant committed:
Merge branch 'master' of https://github.com/thatseant/ip

thatseant committed:
Print Horizontal Line after every Command

thatseant committed:
Add Header Comment for printHorizontalLine
